crates/typst-pdf/src/resources.rs

```diff
@@ -0,0 +1,203 @@
+use std::collections::HashMap;
+use std::hash::Hash;
+
+use ecow::eco_format;
+use pdf_writer::{Dict, Finish, Name, Ref};
+use typst::{text::Font, visualize::Image};
+
+use crate::{
+    color::ColorSpaces, color_font::ColorFontSlice, extg::ExtGState,
+    gradient::PdfGradient, pattern::PdfPattern, AllocGlobalRefs, PdfChunk, Renumber,
+    Resources, WriteResources,
+};
+
+pub struct ResourcesRefs {
+    pub reference: Ref,
+    pub color_fonts: Option<Box<ResourcesRefs>>,
+    pub patterns: Option<Box<ResourcesRefs>>,
+}
+
+impl Default for ResourcesRefs {
+    fn default() -> Self {
+        Self {
+            reference: Ref::new(1),
+            color_fonts: None,
+            patterns: None,
+        }
+    }
+}
+
+impl Renumber for ResourcesRefs {
+    fn renumber(&mut self, old: Ref, new: Ref) {
+        self.reference.renumber(old, new);
+        if let Some(color_fonts) = &mut self.color_fonts {
+            color_fonts.renumber(old, new);
+        }
+        if let Some(patterns) = &mut self.patterns {
+            patterns.renumber(old, new);
+        }
+    }
+}
+
+pub fn alloc_resources_refs(context: &AllocGlobalRefs) -> (PdfChunk, ResourcesRefs) {
+    let mut chunk = PdfChunk::new();
+    fn refs_for(resources: &Resources<()>, chunk: &mut PdfChunk) -> ResourcesRefs {
+        ResourcesRefs {
+            reference: chunk.alloc(),
+            color_fonts: resources
+                .color_fonts
+                .as_ref()
+                .map(|c| Box::new(refs_for(&c.resources, chunk))),
+            patterns: resources
+                .patterns
+                .as_ref()
+                .map(|p| Box::new(refs_for(&p.resources, chunk))),
+        }
+    }
+
+    let refs = refs_for(&context.resources, &mut chunk);
+    (chunk, refs)
+}
+
+/// Write the global resource dictionary that will be referenced by all pages.
+///
+/// We add a reference to this dictionary to each page individually instead of
+/// to the root node of the page tree because using the resource inheritance
+/// feature breaks PDF merging with Apple Preview.
+pub fn write_global_resources(ctx: &WriteResources) -> (PdfChunk, ()) {
```

Comment:
The naming here is a bit confusing since it also writes the local resource dictionaries.